External watch API

aaronlehmann opened the pull request:
Add the first RPC from the external store API: `Watch`.

This RPC supports resuming from a known location, as well as starting the watch stream to only receive future events.

It can optionally return the old version of the object as well as the new one for `Update`s.

cc @dongluochen @aluzzardi

stevvooe commented on api/store.proto:
```diff
@@ -0,0 +1,140 @@
+		StoreActionKind action = 1;
+
+		// Matched object
+		Object object = 2;
```
Can these be `new` and `old` or does that collide with a keyword?

aaronlehmann replied:
I think that would be overly specific to the "update" case. For example, if we're looking at a deletion, `new` wouldn't be the right field name for the object that was deleted.

aaronlehmann commented:
Rebased

dongluochen commented:
Is it easy to make `StoreActionKind` a bitmap combination? For watch request, user can specify all events for service with `StoreActionKindCreate | StoreActionKindUpdate | StoreActionKindRemove`? 

```
StoreActionKindCreate  StoreActionKind = 1
StoreActionKindUpdate  StoreActionKind = 2
StoreActionKindRemove  StoreActionKind = 4
```

aaronlehmann commented:
> Is it easy to make `StoreActionKind` a bitmap combination?

Good idea. Done in the most recent commit. Note that `StoreActionKind` can't be changed because it's part of the wire format, so this introduces a new `WatchActionKind`.

dongluochen commented on api/store.proto:
```diff
@@ -0,0 +1,152 @@
+// Watch. Note that the first item of this stream will always be a WatchMessage
+// with a nil Object, to signal that the stream has started.
+message WatchMessage {
+	message Event {
```
Each `Event` should have a timestamp of commit time to store.

aaronlehmann replied:
Hm, this isn't currently available for historic events. For current `Create` and `Update` events, the individual objects will have the timestamp in `Meta` (but `Delete` events have the old timestamp).

dongluochen replied:
Let's use the `Meta` timestamps.

aaronlehmann replied:
Should we omit it for delete events or include it even though it's the timestamp of the last update?

dongluochen replied:
Each event needs a timestamp. I'll use the last `UpdatedAt` of delete events for now. Another option is to use the current time of machine.

aluzzardi commented on api/store.proto:
```diff
@@ -0,0 +1,152 @@
+syntax = "proto3";
```
I'd add a comment to everything that's not used

aaronlehmann replied:
Not sure I follow. This isn't used yet - it's being added so that events can use it.

aluzzardi commented on api/store.proto:
```diff
@@ -0,0 +1,152 @@
+	// previous version of the object on updates. Note that only live
+	// changes will include the old object (not historical changes
+	// retrieved using ResumeFrom).
+	bool include_old_object = 3;
```
Why do we need for this to be an option rather than the default?

aaronlehmann replied:
@dongluochen was concerned about wasted cycles and bandwidth sending two copies of the object for each change.

dongluochen replied:
For `Event`, it need to include the old object to find the difference. 

If an application just needs the current state and change as trigger, old object is not required. For example, if an application only cares about `create` and `remove` events. it doesn't need old object.

aaronlehmann commented:
Rebased on #2039, and added code generation for all type-specific parts of this code. Now the only code in `watch.go` is a short gRPC stream endpoint. The actual code to build selectors and convert events is all generated.

aaronlehmann commented:
Rebased

dongluochen commented:
LGTM

dongluochen commented:
Ping @docker/core-swarmkit-maintainers.

aaronlehmann commented:
Rebased

diogomonica commented:
@aaronlehmann do we filter secret `data` before we send those objects out?

aaronlehmann commented:
No, this is not doing any redaction. It's the first piece of a store API, that will let external users interact with the data store at a lower level than the control API. I expect in the future there may be use cases where external controllers create and interact with secrets. We could always add redaction now, but might have to reconsider it later.

aluzzardi commented on protobuf/plugin/plugin.proto:
```diff
@@ -4,7 +4,31 @@ package docker.protobuf.plugin;
+}
+
 message StoreObject {
+	required WatchSelectors watch_selectors = 1;
```
is `required` still a thing in proto3?

aaronlehmann replied:
This file doesn't use `proto3` for whatever reason.

aluzzardi commented:
LGTM
